timemachine/cpp/src/random_sampler.hpp

```diff
@@ -0,0 +1,38 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "curand.h"
+#include "device_buffer.hpp"
+
+namespace timemachine {
+
+// An RandomSampler for random sampling from probabilities (np.random.choice(X, replace=True p=probabilities))
+template <typename RealType> class RandomSampler {
+
+private:
+    const int N_; // Max number of values that can be sampled from
+    size_t temp_storage_bytes_;
+
+    DeviceBuffer<RealType> d_rand_;
+    DeviceBuffer<RealType> d_gumbel_;
+
+    std::unique_ptr<DeviceBuffer<char>> d_sort_storage_;
```

Comment:
ugh this `unique_ptr` wrapper is starting to pollute everything - should definitely implement `re-alloc` to avoid having to do this in a future deboggle.

Reply:
Will knock out in a follow up PR, as it is a common use case.